### PR TITLE
New flow registration CLI

### LINKED ISSUE #4233: `prefect register flow` cannot register modules containing relative imports
## Description
Using the prefect CLI to register flows runs into issues when using modules (as in `storage.Module`) instead of scripts (like `storage.Local`, `storage.Github`, etc.) to as flow storage. Using modules enables flows to have tasks moved to other files to encourage task reuse either privately within a project or across the whole agent for more generic tasks.

The problem with using modules this way is that `prefect register flow` currently [only expects to recieve individual files](https://github.com/PrefectHQ/prefect/blob/20cca8a7491079053725526570c5289cf905a751/src/prefect/cli/register.py#L92-L100), so if I have a module like this:

```py
## my_prefect/flows/Module-import/foo.py
from .bar import bar

from prefect import Flow
from prefect.storage.module import Module

script_storage_location = Module("my_prefect.flows.Module-import.foo")

with Flow('my-flow') as flow:
	bar()

if __name__ == '__main__':
	flow.run()
```

```py
## my_prefect/flows/Module-import/bar.py
from prefect import task

@task
def bar():
	pass
```

I can run it in either of these manners:
```
$ python -m my_prefect.flows.Module-import.foo
```
```
$ python my_prefect/flows/Module-import/foo.py
```

But this fails at registration-time because of the way Prefect reads flow files:

```
$ prefect register flow -f my_prefect/flows/Module-import/foo.py -p Module-import
```
```
Traceback (most recent call last):
  File "C:\Program Files\Python\Python38\lib\runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python\Python38\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File ".pyenv\Scripts\prefect.exe\__main__.py", line 9, in <module>

  -- snip --

  File ".pyenv\lib\site-packages\prefect\cli\register.py", line 94, in flow
    flow = extract_flow_from_file(file_path=file_path, flow_name=name)
  File ".pyenv\lib\site-packages\prefect\utilities\storage.py", line 86, in extract_flow_from_file
    exec(contents, exec_vals)
  File "<string>", line 1, in <module>
KeyError: "'__name__' not in globals"
```

While it makes sense to register flows by reading them as standalone files like this for store-as-script backends, this is somewhat of an issue with modules when it comes to relative imports. A (local) agent is perfectly capable of working with this sort of module once it has been registered, assuming the module is on its search path. There simply is no way to register modules that don't turn out to be fully self-contained in a single script without rolling up your own `register flow` command.

## Expected Behavior
The simple fix for this would be to have a `-m` option that uses `extract_flow_from_module` when loading the flow for interpretation instead of `extract_flow_from_file` like `-f` does.

```
$ prefect register flow -m my_prefect.flows.Module-import.foo -p Module-import
```

## Reproduction
See code example in description.

## Environment
I'm using a LocalAgent to run this module, whose working directory is a git working tree that contains the `my_prefect` module, meaning that module is in the agent's search path once it is running.

```
PrefectAgentWorkDir
├── .git
├── .pyenv
├── my_prefect
│   ├── shared
│   └── flows
│       └── Module-import
│           ├── bar.py
│           └── foo.py
├── register_module.py
└── requirements.txt
```

As with a local agent, someone working on the flows is expected to run them from the root of `PrefectAgentWorkDir` in order to have visibility on the same modules.

I know it's a far cry from the nice and clean docker deployments most people probably use, but the flows I'm working on at the moment are ported from preexisting scripts and a number of them are far too complex to neatly fit in a single file.

### PR DESCRIPTION
**Pull request by jcrist**
This PR deprecates the previous `prefect register flow` CLI in favor of a new `prefect register` CLI with lots more functionality.

Features:

- Allows registering multiple flows in a single call. Flows can be specified by `--path` (path to a file or directory containing flows) or by `--module` (an importable python module containing flows). Both options can be specified multiple times in a single call for more flexibility.
- A `--name` flag can be used to subselect flows to only register flows with a specific name (or names). If unspecified, all flows found are registered.
- By default, a flow will only be re-registered if it is *structurally different* than the existing version. This means that small edits to the source of tasks won't require re-registration, and the CLI will automatically detect this to avoid needlessly bumping the version. This can be disabled by passing in `--force`.
- A `--watch` flag for watching for changes in a directory/path/module, and re-registering flows on change. This can be used during development to automatically re-register your flows on save (as needed), or as part of a deployment for users who want to watch and auto-update flows from a specific directory (similar to airflow's DAG directory).

These features simplify auto-registering flows from within CI. For example, with github actions you might add the following step to your CI workflow to auto-register all flows in a `flows/` directory on merge.

```yaml
- name: Register Flows
  if: github.ref == 'refs/heads/master'
  run: prefect register flows --project testing --path flows
```

Fixes #4233.

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Comment by jcrist** [comment]
Demo:

register-cli

File: src/prefect/cli/register.py [hunk 1]
```diff
@@ -1,34 +1,552 @@
...
+    default=False,
+    is_flag=True,
+)
+@click.pass_context
+def register(ctx, project, paths, modules, names, labels, force, watch):
+    """Register one or more flows into a project."""
+    # Since the old command was a subcommand of this, we have to do some
+    # mucking to smoothly deprecate it. Can be removed with `prefect register
+    # flow` is removed.
+    if ctx.invoked_subcommand is not None:
```
Comment (cicdw) [inline src/prefect/cli/register.py:505]: Ah I see what you mean now

**Review by cicdw — CHANGES_REQUESTED** [review]
The most minor request to change a variable name - otherwise this LGTM, and it seems like you've thought through all the edge cases

File: src/prefect/cli/register.py [hunk 2]
```diff
@@ -1,34 +1,552 @@
...
+            out.append(path)
+    return out
+
+
+def load_flows_from_path(path: str) -> "List[prefect.Flow]":
+    """Given a file path, load all flows found in the file"""
+    with open(path, "rb") as fil:
+        contents = fil.read()
+
+    ns = {}
```
[the remark this answers is not in the document]
Comment (jcrist) [reply src/prefect/cli/register.py]: Sure, done.

**Review by cicdw — APPROVED** [review]

**Pull request merged**